File: 1036-rotting-oranges/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        n,m = len(grid),len(grid[0])
        q = deque()
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 2:
                    q.append([i,j,0])
        time = 0
        di = [-1,0,1,0,-1]
        while q:
            x,y,d = q.popleft()
            time = max(time,d)
            for i in range(4):
                dx = di[i]+x
                dy = di[i+1]+y
                if dx < 0 or dy < 0 or dx >= n or dy >= m or grid[dx][dy] != 1:
                    continue
                grid[dx][dy] = 2
                q.append([dx,dy,d+1])
        
        for _ in grid:
            if 1 in _:
                return -1
        return time

File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        queue<pair<int,int>>q;
        int n = grid.size(),m = grid[0].size(),fresh = 0;
        for(int i = 0 ; i < n;i++){
            for(int j = 0 ; j < m ;j++){
                if(grid[i][j] == 2)q.push({i,j});
                if(grid[i][j] == 1)fresh++;
            }
        }
        int minutes = 0;
        vector<int>dir = {-1,0,1,0,-1};
        while(q.size() && fresh){
            minutes++;
            int sz = q.size();
            while(sz--){
                auto [x,y] = q.front();
                q.pop();
                for(int i =0 ; i < 4;i++){
                    int dx = dir[i]+x;
                    int dy = dir[i+1]+y;
                    if(dx >=0 && dy >= 0 && dx < n && dy <m && grid[dx][dy] == 1){
                        q.push({dx,dy});
                        fresh--;
                        grid[dx][dy] = 2;
                    }
                }
            }
        }
        
        return fresh ? -1 : minutes;
    }
};